aggregate/billing_functions/gcp.py

```diff
@@ -0,0 +1,164 @@
+    """
+    From request object, get start and end time if present
+    """
+    start, end = utils.get_start_and_end_from_request(request)
+    main(start, end)
+
+
+def from_pubsub(data=None, _=None):
+    """
+    From pubsub message, get start and end time if present
+    """
+    start, end = utils.get_start_and_end_from_data(data)
+    main(start, end)
+
+
+#################
+#    MIGRATE    #
+#################
+
+
+def migrate_billing_data(start, end, dataset_to_gcp_map) -> int:
+    """
+    Get the billing date in the time period
+    Filter out any rows that aren't in the allowed project ids
+    """
+
+    def get_topic(row):
+        return utils.billing_row_to_topic(row, dataset_to_gcp_map)
+
+    migrate_rows = get_billing_data(start, end)
+
+    if len(migrate_rows) == 0:
+        logger.info('No rows to migrate')
+        return 0
+
+    # Add id and dataset to the row
+    migrate_rows = migrate_rows.drop(columns=['billing_account_id'])
+    migrate_rows.insert(0, 'topic', migrate_rows.apply(get_topic, axis=1))
+    migrate_rows.insert(0, 'id', migrate_rows.apply(billing_row_to_key, axis=1))
+
+    result = utils.insert_dataframe_rows_in_table(
+        utils.GCP_AGGREGATE_DEST_TABLE, migrate_rows
+    )
+
+    return result
+
+
+#################
+#    HELPERS    #
+#################
+
+
+def get_billing_data(start: datetime, end: datetime) -> DataFrame:
+    """
+    Retrieve the billing data from start to end date inclusive
+    Return results as a dataframe
+    """
+
+    _query = f"""
+        SELECT * FROM `{utils.GCP_BILLING_BQ_TABLE}`
+        WHERE export_time >= @start
+            AND export_time <= @end
+    """
+    job_config = bq.QueryJobConfig(
+        query_parameters=[
+            bq.ScalarQueryParameter('start', 'STRING', str(start)),
+            bq.ScalarQueryParameter('end', 'STRING', str(end)),
+        ]
+    )
+
+    migrate_rows = (
+        utils.get_bigquery_client()
+        .query(_query, job_config=job_config)
+        .result()
+        .to_dataframe()
+    )
+
+    return migrate_rows
+
+
+def billing_row_to_key(row) -> str:
+    """Convert a billing row to a hash which will be the row key"""
+    data = tuple(row)
+    identifier = hashlib.md5()
+
+    for item in data:
+        identifier.update(str(item).encode('utf-8'))
+
+    return identifier.hexdigest()
+
+
+def get_dataset_to_gcp_map() -> Dict[str, str]:
+    """Get the server-config from the secret manager"""
+    server_config = json.loads(
+        utils.read_secret(utils.ANALYSIS_RUNNER_PROJECT_ID, 'server-config')
+    )
+    return {v['projectId']: k for k, v in server_config.items()}
+
+
+##############
+#    MAIN    #
+##############
+
+
+def main(start: datetime = None, end: datetime = None) -> int:
+    """Main body function"""
+    interval_iterator = utils.get_date_intervals_for(start, end)
+
+    # Get the dataset to GCP project map
+    dataset_to_gcp_map = get_dataset_to_gcp_map()
+
+    # Migrate the data in batches
```

Comment:
Why are we doing this in batches? (That's not clear from the comment -- it's always good to use comments to explain the reasoning behind something or provide an explanation that's not obvious from the code itself (similarly above "Get the dataset to GCP project map" is exactly what the code directly below says, so it's just adding noise).)

Reply:
I'll clarify here and add the comment. This is because depending on the start-end interval not all of the billing data can be held and manipulated in memory (either locally, but it is also limited for google cloud funcitons). Batching means we can trade storage for time.